src/test/java/com/a2/backend/controller/UserControllerTest.java

```diff
@@ -0,0 +1,205 @@
+import lombok.val;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.web.client.TestRestTemplate;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.HttpStatus;
+import org.springframework.test.annotation.DirtiesContext;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)
+@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.ANY)
+class UserControllerTest {
+
+    @Autowired
+    TestRestTemplate restTemplate;
+
+    private final String baseUrl = "/user";
+
+    private final String nickname = "nickname";
+    private final String email = "some@gmail.com";
+    private final String biography = "bio";
+    private final String password = "password";
+
+    @Test
+    void
+    Test001_GivenAValidUserCreateDTOWhenRequestingPostThenReturnStatusCreatedAndPersistedUserAreReturned() {
+
+        UserCreateDTO userCreateDTO =
+                UserCreateDTO.builder()
+                        .nickname(nickname)
+                        .email(email)
+                        .biography(biography)
+                        .password(password)
+                        .build();
+
+        HttpEntity<UserCreateDTO> request = new HttpEntity<>(userCreateDTO);
+
+        val getResponse = restTemplate.exchange(baseUrl, HttpMethod.POST, request, User.class);
+        assertEquals(HttpStatus.CREATED, getResponse.getStatusCode());
+
+        val persistedUser = getResponse.getBody();
+
+        assertNotNull(persistedUser);
+        assertEquals(nickname, persistedUser.getNickname());
+        assertEquals(email, persistedUser.getEmail());
+        assertEquals(biography, persistedUser.getBiography());
+        assertNotEquals(password, persistedUser.getPassword());
+        assertFalse(persistedUser.isActive());
+    }
+
+    @Test
+    void
+    Test002_GivenAUserCreateDTOWithInvalidNicknameWhenCreatingUserThenBadStatusResponseIsReturned() {
+        String invalidNickname = "not a valid nickname as it is way too long";
+
+        UserCreateDTO userCreateDTO =
+                UserCreateDTO.builder()
+                        .nickname(invalidNickname)
+                        .email(email)
+                        .biography(biography)
+                        .password(password)
+                        .build();
+
+        HttpEntity<UserCreateDTO> request = new HttpEntity<>(userCreateDTO);
+
+        val getResponse = restTemplate.exchange(baseUrl, HttpMethod.POST, request, User.class);
+        assertEquals(HttpStatus.BAD_REQUEST, getResponse.getStatusCode());
```

Comment:
One improvement for the tests would be asserting the error message too

Reply:
Is it ok to add this logic to the controller in order to get the errors generated by the `@Valid` annotation?

`@PostMapping
    public ResponseEntity<User> postNewUser(
            @Valid @RequestBody UserCreateDTO userCreateDTO, Errors errors) {
        if (errors.hasErrors())
            return new ResponseEntity(
                    errors.getAllErrors().get(0).getDefaultMessage(), HttpStatus.BAD_REQUEST);
        val createdUser = userService.createUser(userCreateDTO);
        return ResponseEntity.status(HttpStatus.CREATED).body(createdUser);
    }`

Reply:
You could handle this at the exception handler level. This way, you can write a single handler for every DTO annotated with `@Valid`

Something like:
```java
@ExceptionHandler(MethodArgumentNotValidException.class)
@ResponseStatus(HttpStatus.BAD_REQUEST)
@ResponseBody
public ValidationFailureResponse validationError(MethodArgumentNotValidException ex) {
    BindingResult result = ex.getBindingResult();
    final List<FieldError> fieldErrors = result.getFieldErrors();

    return new ValidationFailureResponse((FieldError[])(fieldErrors.toArray(new FieldError[fieldErrors.size()])));
}
```

Please bear in mind I have extracted this directly from this [stackoverflow post](https://stackoverflow.com/questions/24811643/spring-bean-validation-valid-handling), it might not work "as is" but gives you the general idea.

Overall, we should improve how we are handling exceptions as a whole in the backend. Probably will take some time of the next weekly to focus on this.